src/java/main/org/apache/zookeeper/common/X509Util.java

```diff
@@ -18,64 +18,119 @@
+import java.security.Security;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.security.cert.PKIXBuilderParameters;
+import java.security.cert.X509CertSelector;
+import java.util.Arrays;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import static org.apache.zookeeper.common.X509Exception.KeyManagerException;
-import static org.apache.zookeeper.common.X509Exception.SSLContextException;
-import static org.apache.zookeeper.common.X509Exception.TrustManagerException;
+import org.apache.zookeeper.common.X509Exception.KeyManagerException;
+import org.apache.zookeeper.common.X509Exception.SSLContextException;
+import org.apache.zookeeper.common.X509Exception.TrustManagerException;
 
 /**
  * Utility code for X509 handling
  */
-public class X509Util {
+public abstract class X509Util {
     private static final Logger LOG = LoggerFactory.getLogger(X509Util.class);
 
-    /**
-     * @deprecated Use {@link ZKConfig#SSL_KEYSTORE_LOCATION}
-     *             instead.
-     */
-    @Deprecated
-    public static final String SSL_KEYSTORE_LOCATION = "zookeeper.ssl.keyStore.location";
-    /**
-     * @deprecated Use {@link ZKConfig#SSL_KEYSTORE_PASSWD}
-     *             instead.
-     */
-    @Deprecated
-    public static final String SSL_KEYSTORE_PASSWD = "zookeeper.ssl.keyStore.password";
-    /**
-     * @deprecated Use {@link ZKConfig#SSL_TRUSTSTORE_LOCATION}
-     *             instead.
-     */
-    @Deprecated
-    public static final String SSL_TRUSTSTORE_LOCATION = "zookeeper.ssl.trustStore.location";
-    /**
-     * @deprecated Use {@link ZKConfig#SSL_TRUSTSTORE_PASSWD}
-     *             instead.
-     */
-    @Deprecated
-    public static final String SSL_TRUSTSTORE_PASSWD = "zookeeper.ssl.trustStore.password";
-    /**
-     * @deprecated Use {@link ZKConfig#SSL_AUTHPROVIDER}
-     *             instead.
-     */
-    @Deprecated
-    public static final String SSL_AUTHPROVIDER = "zookeeper.ssl.authProvider";
-
-    public static SSLContext createSSLContext() throws SSLContextException {
-        /**
+    static final String DEFAULT_PROTOCOL = "TLSv1.2";
+
+    private String sslProtocolProperty = getConfigPrefix() + "protocol";
+    private String cipherSuitesProperty = getConfigPrefix() + "ciphersuites";
+    private String sslKeystoreLocationProperty = getConfigPrefix() + "keyStore.location";
+    private String sslKeystorePasswdProperty = getConfigPrefix() + "keyStore.password";
+    private String sslTruststoreLocationProperty = getConfigPrefix() + "trustStore.location";
+    private String sslTruststorePasswdProperty = getConfigPrefix() + "trustStore.password";
+    private String sslHostnameVerificationEnabledProperty = getConfigPrefix() + "hostnameVerification";
+    private String sslCrlEnabledProperty = getConfigPrefix() + "crl";
+    private String sslOcspEnabledProperty = getConfigPrefix() + "ocsp";
+
+    private String[] cipherSuites;
```

Comment:
since we control all the server and client code, it should be safe to just set the cipher suites to a small number of good defaults. I suggest the following 4 suites:

- "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
- "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256"
- "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
- "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"

The ECDHE suites have good performance and security, and AES symmetric encryption is really the only mode that should be considered.

I also did some perf testing and found that the GCM suites perform poorly on Java 8 under load, but perform better than CBC on java 9. In fact, GCM overhead on java 9 was negligible compared to plaintext. So it might make sense to list the suites in different orders based on java version. Something like:

```
  if (java version < 9) {
    cipherSuites = new String[]{
      "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"};
    } else {
      cipherSuites = new String[]{
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"};
    }
```

Reply:
I think this makes a lot of sense and actually was thinking about something similar previously. There might be other clients which use ZooKeeper interface directly, but I think that should not be a problem and safety is more important in this case.

Are you talking about restricting the quorum communication only or set both for server and client connections?

In the latter case we could remove related config properties entirely.

Reply:
Do you think we should follow the same approach with enabled protocols?

Currently there're 2 settings we're dealing with in this patch regarding protocols:
1. a default protocol setting which is configurable and set when we create default SSL context,
2. restriction of enabled protocols which should be set similarly to enabled cipher suites when creating the socket.

The latter one is missing from the code currently, hence the test is failing.

Reply:
I meant quorum only, at least for now. We might need to support more wide variety of suites for clients (I don't plan to start working on client-side TLS until July or August, so I'm not sure yet).

Reply:
@ivmaykov Yep, it does make sense indeed. I've just got an understanding of client-side TLS has been implemented completely different from this and it's built inside Netty.

How about protocol restrictions here?

Reply:
Again, since we control the client and server side of quorum connections, I think it's fine to default to `TLSv1.2`. Someday in the future we'll want to migrate to 1.3, but Java doesn't have support for it yet (indeed, most software does not).

Reply:
Okay. I'll remove the configuration of both enabled cipher suites and protocols. Default protocol remains TLSv1.2, enabled cipher suites will be the list you provided. Thanks.

Reply:
I think it's ok to keep it configurable, just give it good defaults so most people don't need to. For example, if someone wants to use ZK with DSA-signed certs, they would need to specify other cipher suites.

Reply:
Java-specific default cipher suites has been added. I still need to work on protocol restrictions. Do you know what's the correlation between the default protocol set here:
```java
String protocol = System.getProperty(sslProtocolProperty, DEFAULT_PROTOCOL);
try {
    SSLContext sslContext = SSLContext.getInstance(protocol);
    sslContext.init(keyManagers, trustManagers, null);
```
and "enabled protocols" which can be set similar to cipher suites:
```java
sslParameters.setCipherSuites(cipherSuites);
sslParameters.setProtocols(new String[] { "TLS", "SSLv3" ... });
```
Currently, default protocol is configurable. I'm wondering whether it makes sense to add new config property for enabled protocols?

Reply:
I don't think we ever want to use anything earlier than "TLSv1.2", so I wouldn't worry about it too much.

Reply:
I won't touch it then.